Overlay refactor

WSSDude420 opened the pull request:
Refactored the way we hook a bit. 

Also added some better logging for good measure, so we know sooner what is wrong in regards to overlay hook.

All that should be left to do regarding the overlay is to make it compatible with Windows 7, because as of right now, we are not compatible with it.

yamashi commented on src/overlay/Overlay_Hooks.cpp:
```diff
@@ -166,15 +150,13 @@ void Overlay::EarlyHooks(Image* apImage)
-    return Get().m_realPresentD3D12(pSwapChain, SyncInterval, Flags);
+    auto& overlay = Get();
+    
+    if (overlay.InitializeD3D12(pSwapChain)) {
```
line skip

WSSDude420 replied:
corrected

yamashi commented on src/overlay/Overlay_D3D12.cpp:
```diff
@@ -115,23 +214,23 @@ void Overlay::Render(IDXGISwapChain3* pSwapChain)
-    auto& currentFrameContext = m_frameContexts[bufferIndex];
-    currentFrameContext.CommandAllocator->Reset();
+    auto& frameContext = m_frameContexts[bufferIndex];
+    m_pd3dCommandAllocator->Reset();
```
I remember using the same allocator for every frame would make the overlay randomly working or not, are you sure about that?

WSSDude420 replied:
Not 100% sure now tbh. I'll double-check just to be sure after lineskips and such.

But tested all the scenarios I could think of (mixing different overlays running at once with this, only some,...)
Seems like there is no issue with how it is currently in my branch.

WSSDude420 replied:
You were correct, there could be an issue if two threads tried to use it at once, which could technically happen.

So fixed it.

Though it also seems like we should probably have separate command lists for each buffer as well...
Looking at how Reshade does it though, seems like we should be good in command lists regard with single list as we have now.

yamashi commented on src/overlay/Overlay_D3D12.cpp:
```diff
@@ -29,82 +28,182 @@ BOOL CALLBACK EnumWindowsProcMy(HWND hwnd, LPARAM lParam)
+        return reset(this);
+    }
+
+    if (!ImGui_ImplDX12_CreateDeviceObjects()) {
```
line skip

WSSDude420 replied:
corrected

yamashi commented on src/overlay/Overlay_D3D12.cpp:
```diff
@@ -29,82 +28,182 @@ BOOL CALLBACK EnumWindowsProcMy(HWND hwnd, LPARAM lParam)
+    };
+
+    // Window hook (done once)
+    if (m_hWnd == nullptr) {
```
line skip

WSSDude420 replied:
corrected

yamashi commented on src/overlay/Overlay_D3D12.cpp:
```diff
@@ -29,82 +28,182 @@ BOOL CALLBACK EnumWindowsProcMy(HWND hwnd, LPARAM lParam)
+bool Overlay::InitializeD3D12(IDXGISwapChain3* pSwapChain)
 {
-    EnumWindows(EnumWindowsProcMy, reinterpret_cast<LPARAM>(&m_hwnd));
+    static auto checkCmdQueue = [](Overlay* overlay) {
```
line skip

yamashi replied:
also maybe it should be its own function and not a lambda?

WSSDude420 replied:
Don't know honestly if I would make these regular functions. They have meaning only in context of this function, both.

But if you think it would be better, no problem to move both outside. But I really think it is unnecessary.
I also tried to made this lambdas in function format so no functors are in play.

WSSDude420 replied:
for now, corrected line skip here...

gonna leave it as lambda for now till you respond

yamashi commented on src/overlay/Overlay_D3D12.cpp:
```diff
@@ -29,82 +28,182 @@ BOOL CALLBACK EnumWindowsProcMy(HWND hwnd, LPARAM lParam)
+        return true;
+    };
+
+    static auto reset = [](Overlay* overlay) {
```
line skip + same as above

WSSDude420 replied:
for now, corrected line skip here...

gonna leave it as lambda for now till you respond

yamashi commented on src/overlay/Overlay_D3D12.cpp:
```diff
@@ -29,82 +28,182 @@ BOOL CALLBACK EnumWindowsProcMy(HWND hwnd, LPARAM lParam)
+        overlay->m_pd3d12Device = nullptr;
+        overlay->m_pd3dRtvDescHeap = nullptr;
+        overlay->m_pd3dSrvDescHeap = nullptr;
+	    overlay->m_pd3dCommandAllocator = nullptr;
```
indentation

WSSDude420 replied:
corrected

yamashi commented on src/dllmain.cpp:
```diff
@@ -102,6 +102,8 @@ void Shutdown()
 }
 
 BOOL APIENTRY DllMain(HMODULE mod, DWORD ul_reason_for_call, LPVOID) {
+	DisableThreadLibraryCalls(mod);
```
indentation looks wrong

WSSDude420 replied:
no, is right here... problem is that { should be on next line, following your style. 

corrected.

WSSDude420 commented:
Ok, waitting for response regarding those lambdas...

Gonna look up that Command Allocator usage in the mean time.

WSSDude420 commented:
Well, hope it is better now...

Apart from lambdas, all was corrected now.
I also modified a bit log messages in Overlay, particularly ones I added in Overlay::InitializeD3D12(), as I found printing of swap chain argument a bit stupid on second glance.

yamashi commented on src/overlay/Overlay_D3D12.cpp:
```diff
@@ -29,82 +28,186 @@ BOOL CALLBACK EnumWindowsProcMy(HWND hwnd, LPARAM lParam)
+    ImGuiIO& io = ImGui::GetIO();
+    ImGui::StyleColorsDark();
+    io.Fonts->AddFontDefault();
+    io.Fonts->GetTexDataAsRGBA32(&pixels, &width, &height);
```
I don't really see the user of this get?

WSSDude420 replied:
Forgotten leftover from previous code I suppose. I believe I just copied this chunk in particular from former implementation. Will remove it ASAP.

WSSDude420 commented:
Corrected previous issue with latest commit.
